proposals/0000-lambda-layout.md

```diff
@@ -0,0 +1,384 @@
+and allows a new expression to have the same capabilities as function declarations,
+obviating the need for `-XMultiWayIf`, `-XLambdaCase`, and potentially even
+most of function declaration syntax, if a user wishes to use the new expression
+instead (see Example section). This new expression is introduced with the sequence `\ of`. It
+behaves similarly to lambda expressions, except with layout, guards and multiple clauses; for example,
+here is a definition of filter in terms of fix:
+
+```haskell
+filter = fix \of filter' _ []                 -> []
+                 filter' p (x:xs) | p x       -> x : rest
+                                  | otherwise ->     rest
+                   where rest = filter' p xs
+```
+
+The `\of` can be read as "*lambda* comprised *of* the following cases".
+
+By combining the functionality of several features into one feature in
+a way that's consistent with the rest of the language, it enables users who
+wish to use this functionality to work with a simpler and more consistent
+language.
+
+## Proposed Change Specification
+
+A new extension `-XMultiWayLambda` is implemented. Under this extension, a new
+expression is enabled, introduced by the token sequence `\ of`. The whitespace between  `\` and `of`
+is optional and may contain an arbitrary sequence of whitespace characters.
+`\ of` behaves in a way largely similar to `\`, but it is a layout herald, can have multiple
+clauses, and may contain guards (see BNF for details).
+
+Zero clauses are not permitted, as the expression would be ambiguous.
+(See `Alternatives` section for details.)
+This means that with `-XEmptyCase`,
+`-XLambdaCase` still has one (albeit rarely used) construct that cannot be
+replaced by a `\ of`-expression, though
+`\case {}` can (even today) be replaced by `\x -> case x of {}`. This shortcoming
+could potentially
+be addressed in a future proposal, for example by adding absurd patterns, to
+provide a more general solution. Note that the `\case {}` construct only works for matching
+on a single pattern.
+
+Like the existing behavior for alternatives in case- and
+`\case`-expressions, and equations in function declaration syntax, it is
+possible to use `where` clauses within each clause of a `\ of`-expression.
+
+Explicit layout using braces can be used instead of the implicit layout.
+
+As with function declaration equations, all clauses must have the same number of patterns.
+
+Once the [*Binding type variables in lambda-expressions*](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0155-type-lambda.rst) proposal is being implemented,
+with `-XTypeAbstractions`, `\ of`-expressions will also be able to bind type
+variables.
+
+### BNF of changed syntax
+
+*ofexp* → `\` `of` `{` *ofalts* `}`  
+*ofalts* → *ofalt₁* `;` … `;` *ofaltₙ*  
+*ofalt* → [ *apat₁* … *apatₙ* ] `->` *exp* [`where` *decls*]  
+ | [ *apat₁* … *apatₙ* ] *gdpat* [`where` *decls*]
```

Comment:
Note there is technically an ambiguity in the case 0 scrutinees of whether how guards are split among multiple alternatives. But all valid parsings have the same semantics so we can just resolve the ambiguity however we please.